File: src/execution/bin_path.c
#include "environment.h"
#include "libft.h"
#include <fcntl.h>
#include <unistd.h>
#include <stdio.h>
#include "arr_utils.h"
#include "libutils.h"
#include "utils.h"

static uint8_t	find_bin(const char **paths, const char *binprefix
						, char **binpath_buf)
{
	if (!binprefix)
		return (1);
	while (paths && *paths)
	{
		*binpath_buf = ft_strjoin(*paths, binprefix);
		if (binpath_buf && !*binpath_buf)
			return (EXIT_FAILURE);
		if (access(*binpath_buf, F_OK) == 0)
			return (0);
		free_null(binpath_buf);
		paths++;
	}
	return (127);
}

static uint8_t	get_bin(const char **paths, const char *bin, char **binpath_buf)
{
	const char	*bin_prefix = ft_strjoin("/", bin);
	uint8_t		status;

	if (!bin_prefix)
		return (arr_free((char **)paths), 1);
	if (!*bin)
	{
		free((char *)bin_prefix);
		*binpath_buf = ft_strdup("");
		return (arr_free((char **)paths), 127);
	}
	status = find_bin(paths, bin_prefix, binpath_buf);
	arr_free((char **)paths);
	free((char *)bin_prefix);
	if (status == 0 && access(*binpath_buf, X_OK) == -1)
		return (126);
	return (status);
}

uint8_t	set_single(char *const *env, char bin, char **binpath_buf)
{
	if (bin == '/')
		return (eprint("/: Is a directory"), 126);
	if (bin == '.')
		return (eprint(".: filename argument required", *binpath_buf), 2);
	*binpath_buf = get_env(env, "HOME");
	if (!*binpath_buf)
		eprint("/home/minishell: Is a directory");
	else
		eprint("%s: Is a directory", *binpath_buf);
	return (126);
}

/**
 * @brief Get the path to binary
 *
 * @param env shell->env
 * @param bin
 * @param binpath_buf &(token->bin)
 * @return uint8_t 0 on success, 1 on error
 * \return 126 if not executable, 127 if not found
 */
uint8_t	set_binpath(char *const *env, const char *bin, char **binpath_buf)
{
	char		*path;
	const char	**paths;

	if (!bin)
		return (EXIT_FAILURE);
	if (*bin && ft_strchr("~/.", *bin) && !*(bin + 1))
		return (set_single(env, *bin, binpath_buf));
	if (*bin == '.' || *bin == '/')
	{
		*binpath_buf = ft_strdup(bin);
		if (*bin == '.' && *(bin + 1) == '.' && !*(bin + 2))
			return (127);
		if (!*binpath_buf)
			return (EXIT_FAILURE);
		if (access(*binpath_buf, F_OK) == -1)
			return (127);
		if (access(*binpath_buf, X_OK) == -1)
			return (126);
		return (0);
	}
	path = get_env(env, "PATH");
	paths = (const char **)ft_split(path, ':');
	free(path);
	return (get_bin(paths, bin, binpath_buf));
}

File: src/execution/bin_path.rs
use crate::{environment::Env, prelude::*};
use nix::unistd::*;
use std::ffi::{CStr, CString};

fn find_bin(bin_prefix: &str, paths: Vec<String>) -> Option<CString> {
	for path_prefix in paths {
		let bin_path = format!("{}{}", path_prefix, bin_prefix);
		if nix::unistd::access(bin_path.as_str(), AccessFlags::F_OK).is_ok() {
			return Some(CString::new(bin_path).unwrap());
		}
	}
	None
}

fn get_bin(bin: &CStr, paths: Vec<String>) -> (u8, Option<CString>) {
	// prefix bin with /: format will panic if input invalid
	let bin_prefix = format!("{}{}", "/", bin.to_str().unwrap());
	match find_bin(&bin_prefix, paths) {
		Some(bin_path) => {
			if nix::unistd::access(bin_path.as_c_str(), AccessFlags::X_OK).is_err() {
				(126, None)
			} else {
				(0, Some(bin_path))
			}
		}
		_ => (127, None),
	}
}

fn set_single(bin: &CStr, env: &Env) -> (u8, Option<CString>) {
	match bin.to_bytes()[0] {
		b'/' => {
			eprint_msh!("/: Is a directory");
			(126, None)
		}
		b'.' => {
			eprint_msh!(".: filename argument required ");
			(2, None)
		}
		_ => {
			if let Some(home_path) = env.get("HOME") {
				eprint_msh!("{}: Is a directory", home_path);
			} else {
				eprint_msh!("/home/minishell: Is a directory");
			}
			(126, None)
		}
	}
}

pub fn get_path_prefixed(env: &Env, bin: &CStr) -> (u8, Option<CString>) {
	if bin.is_empty() {
		return (127, None);
	}
	let bin_bytes = bin.to_bytes();
	if bin_bytes.len() == 1 {
		if bin_bytes
			.iter()
			.any(|&c| c == b'/' || c == b'.' || c == b'~' || c == b'\0')
		{
			return set_single(bin, env);
		}
		if bin_bytes[0] == b'.' || bin_bytes[0] == b'/' {
			let binpath = CString::from(bin);
			if bin_bytes.len() == 2 && bin_bytes == b".." {
				return (127, None);
			}
			if binpath.is_empty() {
				return (1, None);
			}
			if nix::unistd::access(binpath.as_c_str(), AccessFlags::F_OK).is_err() {
				return (127, None);
			}
			if nix::unistd::access(binpath.as_c_str(), AccessFlags::X_OK).is_err() {
				return (126, None);
			}
			return (0, Some(binpath));
		}
	}
	get_bin(bin, env.get_paths())
}
